### ISSUE TITLE
Pipelining deployments randomly queue for no reason

### ISSUE THREAD

**Issue by jacobbednarz** [zendesk/samson#2176]
Hello!

We recently swapped to a pipelined deployment where we have two stages (canary and production). Canary and production stages are similar however the canary stage only deploys to a tiny subset of our infrastructure whereas production will deploy to the remainder of the infrastructure. Since this swap, we've been seeing random queuing of deployments that never run. Here are the configurations for each stage:

`canary`

![](https://jacobbednarz.s3.amazonaws.com/shares/jxpzg.jpg)

`production`

![](https://jacobbednarz.s3.amazonaws.com/shares/jeykh.jpg)

I've also tried setting the max deployments to one per user but to no avail. Have you got any idea on what it could be or how to debug this further? I'm guessing it's something to do with the cleanup of locks or managing of locks since it's thinking something is deploying but there is nothing there. Appending `debug=1` to the deploys page doesn't show anything.

**Comment by grosser** [comment]
- these 'random' deploys are for the canary stage ?
 - these 'random' deploys are duplicates (deploying the same sha/branch as another deploy) ?

I'd guess there is a webhook that triggers the canary deploy ... then also creates a release ... which triggers another canary deploy ... does removing the `Automatically deploy new releases` help ?

**Comment by jacobbednarz** [comment]
> these 'random' deploys are for the canary stage ?

Sorry, I failed to mention it's only ever `production` that gets held up.

> these 'random' deploys are duplicates (deploying the same sha/branch as another deploy) ?

Nope, usually different commits/branches.

> I'd guess there is a webhook that triggers the canary deploy ... then also creates a release ... which triggers another canary deploy ... does removing the Automatically deploy new releases help ?

Yep, buildkite triggers a release  + deployment. If we turn that off, deployments won't automatically happen.

**Comment by grosser** [comment]
Try turning off the `Automatically deploy new releases`

**Comment by grosser** [comment]
if that solves it then we just need some new logic in the integration controller to never do duplicate deplosy

**Comment by jacobbednarz** [comment]
@grosser I'm slightly confused here with removing that from the canary stage when it's the production stage that is randomly queuing. The canary stage never ends up in a weird state, just the second step of the pipeline.

**Comment by grosser** [comment]
Did you set up a deploy pipeline between canary and production ? ... or are
you manually doing that ?

If there is nothing connecting the 2 stages and no other webhook then
that's really weird :(

**Comment by jacobbednarz** [comment]
I've connected the two with a deploy pipeline. Here is some more config information.

Project

![](https://jacobbednarz.s3.amazonaws.com/shares/f48f4.jpg)

Pipeline config for `canary`

![](https://jacobbednarz.s3.amazonaws.com/shares/4nnut.jpg)

Pipeline config for `production`

![](https://jacobbednarz.s3.amazonaws.com/shares/fgzik.jpg)

Webhooks

![](https://jacobbednarz.s3.amazonaws.com/shares/07deg.jpg)

I'm just at a loss as to what is causing this weirdness because it doesn't happen with a single stage.

**Comment by grosser** [comment]
hmm weird ... so buildkite triggers a release ... which should trigger a deploy to canary and then production ... that sounds about right :/

how about add a webhook mapping for canary and remove the "Automated deploy on release" flag ... that should do the same ... maybe without the bug :/

... otherwise maybe buildkite stores the response of what samson returned ... might include some hints of what it tried to do ... I'll try adding a test for that scenario ... doubt it will find anything though ...

**Comment by jacobbednarz** [comment]
I've swapped to using the webhook for the stage and turned off automatic deployments so let's see how this goes.

Looking at buildkite, I'm not seeing any errors and tailing the logs isn't giving me anything either. Have you got anything else to try in the mean time?

**Comment by grosser** [comment]
Nope, that's about it

**Comment by jacobbednarz** [comment]
After swapping the trigger, it still looks to be an issue :cry: I've tried putting together some test cases this morning however I can't get this to replicate so I've got no idea what is causing this.

**Comment by grosser** [comment]
maybe put a `Rails.logger.error("DEPLOY START" + caller.join("\n"))` into the DeployService so it logs where these deploys are coming from ... the check the logs when it happens again

**Comment by jacobbednarz** [comment]
Have spent a bit more time on this and even removed the pipelining but the random queuing is still present. Where are the stage queue locks kept? This is persisting after application restarts so I don't _think_ it could be suckerpunch.

**Comment by grosser** [comment]
locks are only kept in-memory ... so should be cleared after restart :(

**Comment by jacobbednarz** [comment]
Another odd thing here is that viewing the debug information shows 0 active and 0 queued yet there are 4-5 of these jobs sitting there with the "This deploy is queued, it will be started when previous deploys have finished." message 🤔

**Comment by grosser** [comment]
yeah that message is wrong ... should not say that when it's not in the
queue ... might hint at some bug where the deploy is either created by a
different process (console / different server process / worker) or a bug in
the job queueing ... I can definitely make the error message not misleading

**Comment by jacobbednarz** [comment]
Something else I've found today is that when we try to manually re-deploy, this popup is shown.

![](https://jacobbednarz.s3.amazonaws.com/shares/e7s8r.jpg)

Clicking the details takes us back to a passed CI run 🤔

**Comment by grosser** [comment]
maybe at the time of he check the CI run was still pending ? ... still there on refresh ? ... dig into that more, maybe a issue in commit_status.rb
